### DISCUSSION TITLE
New @protected and/or @library directive

### DISCUSSION THREAD

**Discussion by ugib** [wiremod/wire#2645] [Suggestions]
If you happen to work with a lot of different E2 chips on the same contraption, it might happen that you accidentally override the wrong E2 chip, which means you have to rewire all inputs and outputs. This is especially annoying when working with libraries, i.e. E2s that only contain functions. 

Thus, I suggest either (or both) directives:
`@protected`: Makes it so this E2 can only be overridden/updated with E2s of the same name. 
`@library`: Makes it so the E2 cannot be spawned or be used to overwrite another E2.

Of course, the directive names are merely suggestions. Maybe other terms are more appropriate.

**Comment by vurvdev** [comment]
I don't see the point of this as a directive, maybe can just have a client cvar to not allow replacing E2 chips with different names.

**Reply by ugib** [reply]
The library directive would definitely be useful to mark the code, so you can easily see how it is supposed to be used. This is especially useful when sharing.

**Reply by vurvdev** [reply]
`@protected` should definitely just be a convar. `@library` feels pointless, it just adds yet another directive (it's getting quite bloated) with only the very minimal purpose of erroring for people misunderstanding how your code should be used.

For competent people, add "library" to your `@name` directive, otherwise, not like people who don't know basic E2 would be making use of your library.

In the future if/when directive variables are added you could maybe do this yourself with something akin to python's `__name__`.

**Comment by Fasteroid** [comment]
@ugib Not sure about overwriting another e2, but this will prevent you from spawning one:
```golo
@model models/bull/gates/microcontroller1.mdl

if(entity():model() == "models/bull/gates/microcontroller1.mdl"){
    selfDestruct()
    error("This is a library; #include it in something.")
}
```
I use it in my libraries to notify the user the chip isn't meant to be spawned by itself.
